va/wayland/va_wayland_linux_dmabuf.c

```diff
@@ -0,0 +1,312 @@
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a
+ * copy of this software and associated documentation files (the
+ * "Software"), to deal in the Software without restriction, including
+ * without limitation the rights to use, copy, modify, merge, publish,
+ * distribute, sub license, and/or sell copies of the Software, and to
+ * permit persons to whom the Software is furnished to do so, subject to
+ * the following conditions:
+ *
+ * The above copyright notice and this permission notice (including the
+ * next paragraph) shall be included in all copies or substantial portions
+ * of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+ * OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NON-INFRINGEMENT.
+ * IN NO EVENT SHALL INTEL AND/OR ITS SUPPLIERS BE LIABLE FOR
+ * ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+ * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+ * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+#include "sysdeps.h"
+#include <unistd.h>
+#include <errno.h>
+#include <fcntl.h>
+#include <dlfcn.h>
+#include <sys/stat.h>
+#include <xf86drm.h>
+#include "va_drmcommon.h"
+#include "drm/va_drm_utils.h"
+#include "va_wayland_linux_dmabuf.h"
+#include "va_wayland_private.h"
+#include "linux-dmabuf-unstable-v1-client-protocol.h"
+
+typedef struct va_wayland_linux_dmabuf_context {
+    struct va_wayland_context base;
+    bool                      has_linux_dmabuf;
+    bool                      default_feedback_done;
+} VADisplayContextWaylandLinuxDmabuf;
+
+static void
+feedback_handle_done(
+    void                                *data,
+    struct zwp_linux_dmabuf_feedback_v1 *feedback
+)
+{
+    VADisplayContextP const pDisplayContext = data;
+    struct va_wayland_linux_dmabuf_context *wl_linux_dmabuf_ctx = pDisplayContext->opaque;
+
+    wl_linux_dmabuf_ctx->default_feedback_done = true;
+
+    zwp_linux_dmabuf_feedback_v1_destroy(feedback);
+}
+
+static void
+feedback_handle_format_table(
+    void                                *data,
+    struct zwp_linux_dmabuf_feedback_v1 *feedback,
+    int                                  fd,
+    uint32_t                             size
+)
+{
+    close(fd);
+}
+
+static void
+feedback_handle_main_device(
+    void                                *data,
+    struct zwp_linux_dmabuf_feedback_v1 *feedback,
+    struct wl_array                     *device_array
+)
+{
+    dev_t dev_id;
+    drmDevice *dev;
+    const char *dev_path;
+    VADisplayContextP const pDisplayContext = data;
+    VADriverContextP const ctx = pDisplayContext->pDriverContext;
+    struct drm_state * const drm_state = ctx->drm_state;
+
+    assert(device_array->size == sizeof(dev_id));
+    memcpy(&dev_id, device_array->data, sizeof(dev_id));
+
+    if (drmGetDeviceFromDevId(dev_id, 0, &dev) < 0) {
```

Comment:
only this function need bump libdrm version?
why not use drmGetDevice2 or drmGetDevices2

Reply:
Yeah, if we want to keep the current libdrm version requirement, we can hand-roll our own `drmGetDeviceFromDevId()` implementation (iterating over all devices and checking their device ID). However there is an additional complication: this patch requires wayland-protocols 1.24, and Ubuntu 20.04 only has [wayland-protocols 1.20](https://packages.ubuntu.com/focal/wayland-protocols).

I see two ways to fix this:

- Make the new backend optional, and keep the dependencies in this PR.
- Hand-roll `drmGetDeviceFromDevId()` and also vendor a copy of the [linux-dmabuf XML file](https://gitlab.freedesktop.org/wayland/wayland-protocols/-/blob/main/stable/linux-dmabuf/linux-dmabuf-v1.xml?ref_type=heads).

Which one do you prefer?

Reply:
Bumping to Ubuntu 22.04 is also a reasonable choice, especially since Ubuntu 20.04 will be EOL next year.

Reply:
both looks workable , I prefer 2nd,  because:
first solution actually have 2 separate steps: 1. optional 2. default enabled.   we actually delayed the version bump.  
a potential risk for 2nd solution is: whether the [linux-dmabuf XML file](https://gitlab.freedesktop.org/wayland/wayland-protocols/-/blob/main/stable/linux-dmabuf/linux-dmabuf-v1.xml?ref_type=heads).will be updated and be incompatible in future?

Reply:
It will not. That's kind of the point of it being in "stable".

Reply:
I've updated the patch to hand-roll `drmGetDeviceFromDevId()` and vendor a copy of the linux-dmabuf XML.